### DISCUSSION TITLE
Shroudtopia - v0.0.3

### DISCUSSION THREAD

**Discussion by s0t7x** [s0t7x/shroudtopia#11] [General]
You can download the **Shroudtopia Modloader (Shroudtopia-v0.0.3.zip)** and each example-mod individually.
Place downloaded mods inside a folder called "mods" next to shroudtopia (create one if missing).

_You can also download an example shroudtopia.json configuration which has all mods pre-configured._

_If you would like to instantly get a bundled version of Shroudtopia including all example-mods and experimental stuff you can take a look at the nightly builds in the release section._

**Including Example Mods support Game Version (SVN): 606779**

### What's Changed
* [added Config->modGet*Array. abandoned no_craft_cost, bypass_altar_limit and bypass_world_borders from basics mod as updating it with each update it too stressful and the purpose of basics was to ha…](https://github.com/s0t7x/shroudtopia/pull/10/commits/cbaba5d4c2694232a1b37ffe301d3d8940f6d0ba)
[s0t7x](https://github.com/s0t7x/shroudtopia/commits?author=s0t7x)
* [updated mods to work with 606779 and latest shroudtopia version. added header.gif and adjusted readme.md](https://github.com/s0t7x/shroudtopia/pull/10/commits/8eee08f268c2f573ad48517525f390351f27cb53)
[s0t7x](https://github.com/s0t7x/shroudtopia/commits?author=s0t7x)

**Full Changelog**: https://github.com/s0t7x/shroudtopia/compare/0.2-min...0.0.3

<hr /><em>This discussion was created from the release <a href='https://github.com/s0t7x/shroudtopia/releases/tag/0.0.3'>Shroudtopia - v0.0.3</a>.</em>

**Comment by neverke1** [comment]
Free craft and altar bypass limit was one of the nicest features :c but if it's truly a hassle to maintain then i get getting rid of it for the sake of the rest.
Could create them as separate mods and let the community maintain it if they want idk.

**Reply by s0t7x** [reply]
Free Craft is back <3
https://github.com/s0t7x/shroudtopia/releases/tag/svn-637515-experimental

_Bypassing the altar limit in general is highly unstable and sometimes did corrupt savefiles so I just deprecated it._

**Comment by s0t7x** [comment]
So here is the thing: Shroudtopia is meant to be a mod-loader only. However
the built-in example-mods (which actually are kinda modules to Shroudtopia)
proved to be the main driver for its popularity. Maintaining the mods
currently boils down to reverse updates of the game again and again until I
found some nice way to interop with the application more fluently.

So the mods actually are standalone visual studio solutions and are able to
be maintained separately.

neverke1 ***@***.***> schrieb am Mo., 25. Nov. 2024, 10:29:

**Reply by Egaliterrier** [reply]
I guess that is the whole point though. In order for this project to be more attractive to mod devs than a simple cheat table or standalone injectable dll (both similar to this in that they just patch regions of memory and also needing updates everytime an update invalidates mem patterns/apbs), this needs to do more than load mods - it kind of needs to be more of an actual modding api for the game, offering hooks for functions/calls and all that jazz. Is that in your scope for this project? As you mention, it would need a proper interop library to be developed, which is a tall task, especially for an unknown engine like this. Without or until that exists, I fear modding is probably going to be both sparse and a pain to make and maintain. Aswell as rudimentary in the form of raw memory patches.

**Reply by s0t7x** [reply]
I am eager to get such functionality working. If you take a look at latest source you'll see that I am already preparing funchook, zydis and other re libraries. However reversing the engine takes a very long time and each update can be a huge throwback :(
There is an mostly inactive Enshrouded Modding Discord where folks do share reflections of most game classes but progress is very slow as well.

**Reply by Egaliterrier** [reply]
> I am eager to get such functionality working. If you take a look at latest source you'll see that I am already preparing funchook, zydis and other re libraries. However reversing the engine takes a very long time and each update can be a huge throwback :(
> There is an mostly inactive Enshrouded Modding Discord where folks do share reflections of most game classes but progress is very slow as well.

Didn’t see that, so cool you’re starting preliminary work on it. Once it gets going, how about calling the project “Unshrouded”? That would fit the nature of it perfectly:D True! And it’s especially time consuming when there aren’t more devs actively working on it. Yeah, I already lurk there, I wish it was more active though.
Was thinking about ways to attract more hands to the table. People behind projects like ue4ss or bepinex/melonloader and their backend tooling (il2cppinterop, patternsleuth etc.) would be amazing to have assist in this kind of work, but they probably have their hands full, or they might just be burnt out and wanting to tackle something new. And I am sure there are people who would love this kind of challenge, they just have to know about it first. I might post something on some modding/reversing discords, get more eyes on it.

**Comment by s0t7x** [comment]
Not saying it's not hard work but looking especially on ue4ss and bepinex
isn't a fair comparison. Bepinex tackles c# runtime which is a JIT language
and therefore provides lots of reflection information on runtime. UE is an
Open-Source engine which removes a lot of reversing work. I would love to
get a glimpse of what the Enshrouded source looks like but as long as no
one leaks something it's gonna be me versus ASM :(

Egaliterrier ***@***.***> schrieb am Do., 28. Nov. 2024,
13:46:

**Reply by Egaliterrier** [reply]
Ahh, right, I hadn’t thought of how much really gets handed to those projects compared to a novel engine like this, obviously closed source is a much tougher nut to crack. Welp, I guess until someone leaks a nutcracker of a pdb, or they provide some rudimentary modding api, happy ghidra-ing!:)

**Comment by lordfattimus** [comment]
Hi!

Can you provide instructions for how to use the whitelist mod? Thanks!

**Comment by s0t7x** [comment]
At it to the shroudtopia.json mods section like this:
"Whitelist": {
"active": true,
"steamIds": [ ... ]
}

You can take a look at the shroudtopia.json of this release.

Put the steamIds (steamID64 (Dec)) you want to whitelist comma-seperated
into the array.

You can use www.steamidfinder.com to lookup SteamId64 in decimal format.

lordfattimus ***@***.***> schrieb am So., 15. Dez. 2024,
21:20:

**Reply by lordfattimus** [reply]
Thanks! I found this in the example and tried it, but doesn't seem to work. I tried with commas and semicolons with and without spaces, but it gives error that server is full or unavailable. It works if I disable the mod. Here's how my shroudtopia.json looks: 

`{
    "active": true,
    "bootDelay": 3000,
    "enableLogging": true,
    "logLevel": "INFO",
    "mods": {
		"Whitelist": {
			"active": true,
			"steamIDs": [xxxxxxxx, xxxxxxxx]
		}
	},
    "updateDelay": 500
}
`

Oh and yes I'm sure my steamID is right and in Dec format

**Reply by s0t7x** [reply]
Sorry iam late but take a look at your config. The property key for your steamId array is: steamIDs but should be steamIds with a lower case d

**Reply by lordfattimus** [reply]
Thanks, but unfortunately still not working. It crashes the server on the "Activating mod: Whitelist" step per the log. There's no error in shroudtopia.log or in the server logs files, just stops after "Activating mod: Whitelist". I copied your shroudtopia.json to make sure I didn't change anything else. 

This happens if I set whitelist active to true, regardless of the rest of the settings in the mod section. The server will run if set to false, and if I change to true while the server is running that also crashes once it updates. 

Is this just a new issue now because of recent hotfixes or something?

**Comment by s0t7x** [comment]
Whitelist mod is not up-to-date yet. See the release description.

lordfattimus ***@***.***> schrieb am Sa., 8. Feb. 2025, 22:14:

**Reply by lordfattimus** [reply]
Ah, no worries at all. I will keep an eye for any updates. Thanks.
